lib/SPIRV/SPIRVToOCL.cpp

```diff
@@ -247,51 +247,183 @@ void SPIRVToOCL::visitCallSPRIVImageQuerySize(CallInst *CI) {
-    std::string GroupOp = "";
-    switch (GO) {
-    case GroupOperationReduce:
-      GroupOp = "reduce";
-      break;
-    case GroupOperationInclusiveScan:
-      GroupOp = "scan_inclusive";
-      break;
-    case GroupOperationExclusiveScan:
-      GroupOp = "scan_exclusive";
-      break;
-    default:
-      assert(!"Unsupported group operation");
-      break;
-    }
-    DemangledName = Prefix + kSPIRVName::GroupPrefix + GroupOp + '_' + Op.str();
+  std::string GroupOp;
+  std::string GroupPrefix = kSPIRVName::GroupNonUniformPrefix;
+  auto GO = getArgAs<spv::GroupOperation>(CI, 1);
+  switch (GO) {
+  case GroupOperationReduce:
+    GroupOp = "reduce";
+    break;
+  case GroupOperationInclusiveScan:
+    GroupOp = "scan_inclusive";
+    break;
+  case GroupOperationExclusiveScan:
+    GroupOp = "scan_exclusive";
+    break;
+  case GroupOperationClusteredReduce:
+    GroupOp = "clustered_reduce";
+    // OpenCL clustered builtin has no non_uniform prefix, ex.
+    // sub_group_reduce_clustered_logical_and
+    GroupPrefix = kSPIRVName::GroupPrefix;
+    break;
+  default:
+    llvm_unreachable("Unsupported group operation!");
+    break;
+  }
+
+  return Prefix + GroupPrefix + GroupOp + "_" + Op;
+}
+
+std::string SPIRVToOCL::getBallotBuiltinName(CallInst *CI, Op OC) {
+  assert((OC == OpGroupNonUniformBallotBitCount) &&
+         "Not inteded to handle other opcodes than "
+         "OpGroupNonUniformBallotBitCount!");
+  std::string Prefix = getGroupBuiltinPrefix(CI);
+  assert(
+      (Prefix == kOCLBuiltinName::SubPrefix) &&
+      "Workgroup scope is not supported for OpGroupNonUniformBallotBitCount");
+  std::string GroupOp;
+  auto GO = getArgAs<spv::GroupOperation>(CI, 1);
+  switch (GO) {
+  case GroupOperationReduce:
+    GroupOp = "bit_count";
+    break;
+  case GroupOperationInclusiveScan:
+    GroupOp = "inclusive_scan";
+    break;
+  case GroupOperationExclusiveScan:
+    GroupOp = "exclusive_scan";
+    break;
+  default:
+    llvm_unreachable("Unsupported group operation!");
+    break;
   }
+  return Prefix + kSPIRVName::GroupPrefix + "ballot_" + GroupOp;
+}
+
+std::string SPIRVToOCL::groupOCToOCLBuiltinName(CallInst *CI, Op OC) {
+  auto FuncName = OCLSPIRVBuiltinMap::rmap(OC);
+  assert(FuncName.find(kSPIRVName::GroupPrefix) == 0);
+
+  if (!hasGroupOperation(OC)) {
+    /// Transform OpenCL group builtin function names from group_
+    /// to work_group_ and sub_group_.
+    FuncName = getGroupBuiltinPrefix(CI) + FuncName;
+  } else { // Opcodes with group operation parameter
+    if (isUniformArithmeticOpCode(OC))
+      FuncName = getUniformArithmeticBuiltinName(CI, OC);
+    else if (isNonUniformArithmeticOpCode(OC))
+      FuncName = getNonUniformArithmeticBuiltinName(CI, OC);
+    else if (OC == OpGroupNonUniformBallotBitCount)
+      FuncName = getBallotBuiltinName(CI, OC);
+    else
+      llvm_unreachable("Unsupported opcode!");
+  }
+  return FuncName;
+}
+
+bool SPIRVToOCL::extendRetTyToi32(Op OC) {
```

Comment:
```suggestion
bool SPIRVToOCL::needToExtendRetTyToi32(Op OC) {
```

Also, it is probably better to make it just a `static` function instead of class member function

Reply:
Done.